packages/grafana-runtime/src/utils/useFavoriteDatasources.ts

```diff
@@ -0,0 +1,120 @@
+ */
+export function useFavoriteDatasources(): {
+  favoriteDatasources: string[];
+  initialFavoriteDataSources: string[];
+  addFavoriteDatasource: (ds: DataSourceInstanceSettings) => void;
+  removeFavoriteDatasource: (ds: DataSourceInstanceSettings) => void;
+  isFavoriteDatasource: (dsUid: string) => boolean;
+  toggleFavoriteDatasource: (ds: DataSourceInstanceSettings) => void;
+} {
+  const [userStorage] = useState(() => new UserStorage('grafana-runtime'));
+  const [favoriteDatasources, setFavoriteDatasources] = useState<string[]>([]);
+  const [initialFavoriteDataSources, setInitialFavoriteDataSources] = useState<string[]>([]);
+
+  // Load favorites from storage on mount
+  useEffect(() => {
+    const loadFavorites = async () => {
+      try {
+        const stored = await userStorage.getItem(FAVORITE_DATASOURCES_KEY);
+        if (stored) {
+          const parsed = JSON.parse(stored);
+          if (Array.isArray(parsed)) {
+            setFavoriteDatasources(parsed);
+            setInitialFavoriteDataSources(parsed);
+          }
+        }
+      } catch (error) {
+        console.warn('Failed to load favorite data sources:', error);
+        setFavoriteDatasources([]);
+      }
+    };
+
+    loadFavorites();
+  }, [userStorage]);
+
+  // Helper function to save favorites to storage
+  const saveFavorites = useCallback(
+    async (newFavorites: string[]) => {
+      try {
+        await userStorage.setItem(FAVORITE_DATASOURCES_KEY, JSON.stringify(newFavorites));
+        setFavoriteDatasources(newFavorites);
+      } catch (error) {
+        console.warn('Failed to save favorite data sources:', error);
+      }
+    },
+    [userStorage]
+  );
+
+  const addFavoriteDatasource = useCallback(
+    (ds: DataSourceInstanceSettings) => {
+      if (ds.meta.builtIn) {
+        // Prevent storing built-in datasources (-- Grafana --, -- Mixed --, -- Dashboard --)
+        return;
+      }
+
+      if (!favoriteDatasources.includes(ds.uid)) {
+        const newFavorites = [...favoriteDatasources, ds.uid];
+        saveFavorites(newFavorites);
+      }
+    },
+    [favoriteDatasources, saveFavorites]
+  );
+
+  const removeFavoriteDatasource = useCallback(
+    (ds: DataSourceInstanceSettings) => {
+      const newFavorites = favoriteDatasources.filter((uid) => uid !== ds.uid);
+      if (newFavorites.length !== favoriteDatasources.length) {
+        saveFavorites(newFavorites);
+      }
+    },
+    [favoriteDatasources, saveFavorites]
+  );
+
+  const isFavoriteDatasource = useCallback(
+    (dsUid: string) => {
+      return favoriteDatasources.includes(dsUid);
+    },
+    [favoriteDatasources]
+  );
+
+  const toggleFavoriteDatasource = useCallback(
+    (ds: DataSourceInstanceSettings) => {
+      if (isFavoriteDatasource(ds.uid)) {
+        removeFavoriteDatasource(ds);
+      } else {
+        addFavoriteDatasource(ds);
+      }
+    },
+    [isFavoriteDatasource, addFavoriteDatasource, removeFavoriteDatasource]
+  );
```

Comment:
I'm not sure an explicit `toggle` function is useful.

If a user clicks a button `[Unfav data source]`, the user is not asking to _toggle_ it. The only valid action is to remove it.